### PR TITLE
feat: add `attachment.sidecar` publish option

### PR DESCRIPTION
**Pull request by aleclarson**
The term "sidecar" is what Threads uses to represent a group of images and/or videos that share the same post.

### Upload ID collisions

I also added logic to make sure that upload IDs never collide (i.e. if two publish calls happen at the same time).

### Deprecations

As part of this PR, I propose we deprecate the `url` and `image` publish options, replacing them with a single `attachment` option which can be `{ url }`, `{ image }`, or `{ sidecar }`.

**Before**
```ts
await threads.publish({
  text: "Hello",
  url: options.url,
  image: options.image,
  sidecar: options.sidecar,
})
```

**After**
```ts
await threads.publish({
  text: "Hello",
  attachment: options.attachment,
})
```

This means we can…
- better enforce one property only (i.e. `url` and `image` cannot both be defined)
- and it gives users a useful `PostAttachment` type

The example below demonstrates the strict enforcement:

https://www.typescriptlang.org/play?useUnknownInCatchVariables=false&ssl=9&ssc=1&pln=10&pc=1#code/C4TwDgpgBACgrgIwDYEsDOALA8mYKD2AdmgGoCMUAvFAN4BQUjUwEAHsAFxRrABOKhAOYBuOgF8oAMigAKGlDi8kAfi49+QqBIA+tKCgC2AQ0ERV3PgMFaAlHToATCAGMkR3tABmcQs7xEoMERUTAB9ADcyGXxcAmIueGR0bFiiUjIbLnD8FAd7UEgoABkBAGsAQWBgI2cMAwhCYCo9RSQ1S00xOgLoAEljU0rq2vrG5vlDEwh2jWsuugEWXk8a6ESQlP9iEgAmWgYmFnYZq1EmKCMqmrqG4HMAZUs-AFVCOIAeEsIKq5HbqF0-SmQ2uo2AAD5xPYnK53F4fH44oFgskIjtoql4rAUZgcFtSDtMlBsrl7N5fFsoGwjAYwEgIBEojZ9ucPMBFIRkUkwpE5AdzowjpwoAAiAASECQSHwUAA7vglA4RQAafkCgD06rV501UGeaAgvH0aCgACoAHJYAAqprl7kIEAcFwQ+DgTRdwAwUAABpNTN6LoQnd7WgG2Og8EIAHTapia2OMP3TKAAcgwVTAaA4mpQUcMgkUUec+AM6oAouVCHBemgDFGAFaQQQp5VQBMKJRcNMZrOawT4fCCelFksptvnMR2ebkxEBam0+lomTM+isiDs3icoLcjBL1cCwVsYXiyXSuUKpBK1UHxjxm9QXX6w3Gu2bx3MDCXUVJkWBp0i1pf2cIxCEIfB3XAr1wx4KMoAATSMEAAEJ2zvG9LmGG5Gi4eR23OJMu3TYBM2zdVc3zQti1LCsqxrOtGwgZtWzwphWkIntSP7Qdhyosd2zEa8mEnKFunAaB7kMOkUE8EB3itcFxgkaR5AAbRgfROVKCAQHwTwoCtABdLgrTUgytFEUTCkefgXjeIg5IU6gAGESwQAQIFeD55KpdgGgcE0BE8Z9ngOZR9J8lggxNECQDVMKJIXaTZKtKQ9BUgBpDSoDLVhXDgJx3i0nS9OeVsit0-TwQM8wfCcTx3KdMRIXOLgHXCQ0DlaiB2t4CyeigFyDDch1PPs7zqBgFBnFKOSyu0irvMkA4YHcPAjCQd4ZBSo8-OiwgQCgMKZFYYzmUoBSYqgLqeuZHaotkE6NKCo1njOi79sOvUrqgNrDXBUQgA

Finally, I propose that we deprecate the exported types whose name begins with `ThreadsAPI` (except for `ThreadsAPIOptions`), since this isn't a naming convention we apply evenly. Therefore, `ThreadsAPIImage` becomes `ImageDefinition` (as `Image` alone seems too vague?). If we wanted to, we could put all types in an `export namespace ThreadsAPI` block instead. But whatever we do, let's keep it consistent. 👍

### CODE REVIEW

File: threads-api/src/threads-api.ts [hunk 1]
```diff
@@ -148,23 +148,50 @@ export type ThreadsAPIOptions = {
...
   mentioned_only: 2,
-};
+} as const;
+
+export type PostReplyControl = keyof typeof REPLY_CONTROL_OPTIONS;
+
+/** @deprecated Use `PostReplyControl` instead. */
+export type ThreadsAPIPostReplyControl = PostReplyControl;
+
+export type ImageDefinition = { path: string } | { type: string; data: Buffer };
```
Comment (aleclarson) [inline threads-api/src/threads-api.ts:163]: Could also be `PublishedImage` or `ImageUpload`

File: threads-api/src/threads-api.ts [hunk 2]
```diff
@@ -1141,3 +1206,14 @@ export class ThreadsAPI {
...
+type Simplify<T> = {} & { [P in keyof T]: T[P] };
+
+type StrictUnion<T> = CombineUnion<T> extends infer U
+  ? T extends any
+    ? Simplify<T & { [K in Exclude<keyof U, keyof T>]?: undefined }>
+    : never
+  : never;
+
+type CombineUnion<T> = Pick<T, keyof T> &
+  Partial<(T extends any ? (x: T) => any : never) extends (x: infer U) => any ? U : never>;
```
Comment (junhoyeo) [inline threads-api/src/threads-api.ts]: Have a feeling that we'll use this handy! Can we move this to `types/utils.ts` or some other separate file besides `threads-api.ts`?

File: threads-api/src/threads-api.ts [hunk 3]
```diff
@@ -979,29 +1023,48 @@ export class ThreadsAPI {
...
     }
 
-    const now = new Date();
-    const timezoneOffset = -now.getTimezoneOffset() * 60;
-
     let data: any = {
+      ...this._createUploadMetadata(),
       text_post_app_info: {
-        reply_control: REPLY_CONTROL_IDS[options.replyControl ?? 'everyone'],
+        reply_control: REPLY_CONTROL_OPTIONS[options.replyControl ?? 'everyone'],
```
Comment (junhoyeo) [inline threads-api/src/threads-api.ts:1029]: 👍👍

File: threads-api/src/threads-api.ts [hunk 4]
```diff
@@ -148,23 +149,50 @@ export type ThreadsAPIOptions = {
...
-      quotedPostID?: string;
-    } & ({ url?: string } | { image?: string | ThreadsAPIImage });
+/** @deprecated Use `PostReplyControl` instead. */
+export type ThreadsAPIPostReplyControl = PostReplyControl;
 
-export type ThreadsAPIImage = { path: string } | { type: string; data: Buffer };
+export type ImageDefinition = { path: string } | { type: string; data: Buffer };
+
+/** @deprecated Use `ImageDefinition` instead. */
+export type ThreadsAPIImage = ImageDefinition;
```
Comment (junhoyeo) [inline threads-api/src/threads-api.ts:166]: Nice tagging! 👍👍

**Review by junhoyeo — APPROVED** [review]
Yo @aleclarson. Thank you so much for your work! Sidecar integration is truly awesome. 😄 

> Finally, I propose that we deprecate the exported types whose name begins with ThreadsAPI (except for ThreadsAPIOptions), since this isn't a naming convention we apply evenly. Therefore, ThreadsAPIImage becomes ImageDefinition (as Image alone seems too vague?). If we wanted to, we could put all types in an export namespace ThreadsAPI block instead. But whatever we do, let's keep it consistent. 👍

You have a point: I hadn't had a straightforward naming convention when I first launched this package, so I agree with your every word!
But one thing that catches my mind is that I added the `ThreadsAPI` prefix for types that are exposed to the user related to the `ThreadsAPI` class (just like `ThreadsAPIOptions`).
Since `ImageDefinition` (or possible name: `ThreadsAPIImageDefinition`) is also an interface that is not used in threads directly and is for interacting with our API client, I think it should still have the `ThreadsAPI` prefix.

Wdyt? I'm open to your ideas about the naming part 👍👍

**Comment by aleclarson** [comment]
I'm onboard with namespacing any unofficial data structure with `ThreadsAPI`. I propose we go with a `namespace ThreadsAPI` approach, as it's more readable, requires less importing, and I don't see a reason not to take advantage of TypeScript's namespace feature.

**Comment by junhoyeo** [comment]
> I'm onboard with namespacing any unofficial data structure with `ThreadsAPI`. I propose we go with a `namespace ThreadsAPI` approach, as it's more readable, requires less importing, and I don't see a reason not to take advantage of TypeScript's namespace feature.

Looks great to me! Thanks, I'll merge. 👍

**Review by junhoyeo — APPROVED** [review]
Thank you for your meticulous work! @aleclarson

**Pull request merged**